src/SAWScript/MGU.hs

```diff
@@ -441,28 +451,42 @@ resolveCurrentTypedefs t = do
   return $ instantiate s t
 
 -- Get the unification vars that are used in the current variable typing
--- environment.
+-- and named type environments.
 --
 -- FIXME: This function may miss type variables that occur in the type
 -- of a binding that has been shadowed by another value with the same
 -- name. This could potentially cause a run-time type error if the
 -- type of a local function gets generalized too much. We can probably
 -- wait to fix it until someone finds a sawscript program that breaks.
-unifyVarsInEnv :: TI (M.Map TypeIndex Pos)
-unifyVarsInEnv = do
-  env <- TI $ asks varEnv
-  let ss = M.elems env
-  ss' <- mapM applyCurrentSubst ss
-  return $ unifyVars ss'
-
--- Get the named typedef vars that occur in the current variable typing
--- environment.
-namedVarsInEnv :: TI (M.Map Name Pos)
-namedVarsInEnv = do
-  env <- TI $ asks varEnv
-  let ss = M.elems env
-  ss' <- mapM applyCurrentSubst ss
-  return $ namedVars ss'
+--
+-- dholland 20241220: I don't think that's a problem. If there's a
+-- loose unification var somewhere that's been shadowed to the point
+-- where it's not accessible, we can't have accessed it in order to
+-- generate a reference to it in the current block. If it is somewhere
+-- accessible, we'll find it there. This might have broken in the past
+-- when it didn't search the named type environment, but that leak has
+-- been corrected.
+--
+-- Note that we apply the current substitution first. This means that
+-- the caller must also apply the current substitution before reasoning
+-- about what unification vars do and don't appear.
+--
+-- Returns a map of the index number to the occurrence position.
+unifyVarsInEnvs :: TI (M.Map TypeIndex Pos)
+unifyVarsInEnvs = do
+  venv <- TI $ asks varEnv
+  tenv <- TI $ asks tyEnv
+  vtys <- mapM applyCurrentSubst $ M.elems venv
+  ttys <- mapM applyCurrentSubst $ M.elems tenv
+  return $ M.unionWith choosePos (unifyVars vtys) (unifyVars ttys)
+
+-- Get the named type vars that occur as keys in the current type name
+-- environment. Returns a map to unit rather than a set for the
+-- caller's convenience.
+namedVarDefinitions :: TI (M.Map Name ())
+namedVarDefinitions = do
+   env <- TI $ asks tyEnv
+   return $ M.map (\_ -> ()) env
```

Comment:
Similarly, we could simplify this definition by just using [`keysSet`](https://hackage.haskell.org/package/containers-0.7/docs/Data-Map-Lazy.html#v:keysSet) instead.